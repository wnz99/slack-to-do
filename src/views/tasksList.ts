import {
  Actions,
  Button,
  Divider,
  EasyPaginator,
  Section,
} from 'slack-block-builder'

import { renderDate } from '../lib/utils'
import { TasksList, UserActions } from '../types'

export const tasksList = ({ tasks, perPage, page }: TasksList) =>
  EasyPaginator({
    perPage,
    items: tasks,
    page: page || 1,
    actionId: ({ page, offset }) =>
      JSON.stringify({ action: UserActions.UpdateTasksList, page, offset }),
    blocksForEach: ({ item }) => [
      Divider(),
      Section({
        text: `*${item.title}* ${item.isCompleted ? ':white_check_mark:' : ''}`,
      }),
      Section({
        text: `Due Date: ${renderDate(item.dueDate)}`,
      }),
      Actions().elements(
        Button({ text: 'View' })
          .actionId(UserActions.ViewTask)
          .value(
            JSON.stringify({
              taskId: item.id.toString(),
            })
          ),
        Button({ text: item.isCompleted ? 'Open' : 'Complete' })
          .actionId(UserActions.SetTaskComplete)
          .value(
            JSON.stringify({
              taskId: item.id.toString(),
              value: !item.isCompleted,
            })
          )
      ),
    ],
  }).getBlocks()
