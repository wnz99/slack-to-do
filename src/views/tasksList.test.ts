import { tasksList } from './tasksList'

const tasks = [
  {
    id: 'M4ISI5reY86oEQ36Xhiw',
    createdAt: '1658763404',
    title: '6',
    isCompleted: false,
    dueDate: undefined,
  },
  {
    id: 'Qcks3amXXe3lNHe7J2vk',
    isCompleted: false,
    title: '6',
    createdAt: '1658763382',
    dueDate: undefined,
  },
]

describe('tasksList', () => {
  it('should render without errors', () => {
    const blocks = tasksList({ tasks, perPage: 5, page: 1 })

    expect(blocks).toMatchSnapshot()
  })
})
