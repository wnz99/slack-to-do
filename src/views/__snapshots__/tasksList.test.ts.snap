// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tasksList should render without errors 1`] = `
Array [
  DividerBuilder {
    "props": Object {},
  },
  SectionBuilder {
    "props": Object {
      "text": "*6* ",
    },
  },
  SectionBuilder {
    "props": Object {
      "text": "Due Date: none",
    },
  },
  ActionsBuilder {
    "props": Object {
      "elements": Array [
        ButtonBuilder {
          "props": Object {
            "actionId": "view_task",
            "text": "View",
            "value": "{\\"taskId\\":\\"M4ISI5reY86oEQ36Xhiw\\"}",
          },
        },
        ButtonBuilder {
          "props": Object {
            "actionId": "set_task_complete",
            "text": "Complete",
            "value": "{\\"taskId\\":\\"M4ISI5reY86oEQ36Xhiw\\",\\"value\\":true}",
          },
        },
      ],
    },
  },
  DividerBuilder {
    "props": Object {},
  },
  SectionBuilder {
    "props": Object {
      "text": "*6* ",
    },
  },
  SectionBuilder {
    "props": Object {
      "text": "Due Date: none",
    },
  },
  ActionsBuilder {
    "props": Object {
      "elements": Array [
        ButtonBuilder {
          "props": Object {
            "actionId": "view_task",
            "text": "View",
            "value": "{\\"taskId\\":\\"Qcks3amXXe3lNHe7J2vk\\"}",
          },
        },
        ButtonBuilder {
          "props": Object {
            "actionId": "set_task_complete",
            "text": "Complete",
            "value": "{\\"taskId\\":\\"Qcks3amXXe3lNHe7J2vk\\",\\"value\\":true}",
          },
        },
      ],
    },
  },
]
`;
